warp10/src/main/java/io/warp10/script/MemoryWarpScriptStack.java

```diff
@@ -594,308 +595,321 @@ public void exec(String line) throws WarpScriptException {
-          }
-          
-          func = null != func ? func : WarpScriptLib.getFunction(stmt);
+            func = null != func ? func : defined.get(stmt);
 
-          if (null == func) {
-            throw new WarpScriptException("Unknown function '" + stmt + "'");
-          }
+            if (null != func && Boolean.FALSE.equals(getAttribute(WarpScriptStack.ATTRIBUTE_ALLOW_REDEFINED))) {
+              throw new WarpScriptException("Disallowed redefined function '" + stmt + "'.");
+            }
 
-          Map<String,String> labels = new HashMap<String,String>();
-          labels.put(SensisionConstants.SENSISION_LABEL_FUNCTION, stmt);
-          
-          long nano = System.nanoTime();
-          
-          try {
-            if (func instanceof WarpScriptStackFunction && macros.isEmpty()) {
-              //
-              // Function is an WarpScriptStackFunction, call it on this stack
-              //
-              
-              WarpScriptStackFunction esf = (WarpScriptStackFunction) func;
+            func = null != func ? func : WarpScriptLib.getFunction(stmt);
 
-              esf.apply(this);
-            } else {
-              //
-              // Push any other type of function onto the stack
-              //
-              if (macros.isEmpty()) {
-                push(func);
+            if (null == func) {
+              throw new WarpScriptException("Unknown function '" + stmt + "'");
+            }
+
+            Map<String,String> labels = new HashMap<String,String>();
+            labels.put(SensisionConstants.SENSISION_LABEL_FUNCTION, stmt);
+
+            long nano = System.nanoTime();
+
+            try {
+              if (func instanceof WarpScriptStackFunction && macros.isEmpty()) {
+                //
+                // Function is an WarpScriptStackFunction, call it on this stack
+                //
+
+                WarpScriptStackFunction esf = (WarpScriptStackFunction) func;
+
+                esf.apply(this);
               } else {
-                macros.get(0).add(func);
+                //
+                // Push any other type of function onto the stack
+                //
+                if (macros.isEmpty()) {
+                  push(func);
+                } else {
+                  macros.get(0).add(func);
+                }
               }
-            }          
-          } finally {
-            Sensision.update(SensisionConstants.SENSISION_CLASS_WARPSCRIPT_FUNCTION_COUNT, labels, 1);
-            Sensision.update(SensisionConstants.SENSISION_CLASS_WARPSCRIPT_FUNCTION_TIME_US, labels, (System.nanoTime() - nano) / 1000L);
+            } finally {
+              Sensision.update(SensisionConstants.SENSISION_CLASS_WARPSCRIPT_FUNCTION_COUNT, labels, 1);
+              Sensision.update(SensisionConstants.SENSISION_CLASS_WARPSCRIPT_FUNCTION_TIME_US, labels, (System.nanoTime() - nano) / 1000L);
+            }
+          }
+        } catch (WarpScriptException e) {
```

Comment:
You cannot catch WarpScriptException without previously catching WarpScriptATCException, otherwise your throw clause will break any ongoing asynchronous transfer of control, you need to add a catch clause for at least WarpScriptATCException with a body that rethrows it as is.